In [3]:
import warnings
warnings.filterwarnings('ignore')

In [4]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

# Loading our df

In [5]:
df=pd.read_csv('Property_with_Feature_Engineering.csv')
df

property_id  location_id  \
0            347795            8   
1            482892           48   
2            555962           75   
3            562843         3821   
4            686990         3522   
...             ...          ...   
191388     17468383          174   
191389     17468384          174   
191390     17468482          167   
191391     17468586          339   
191392     17468660         3421   

                                                 page_url  property_type  \
0       https://www.zameen.com/Property/lahore_model_t...          House   
1       https://www.zameen.com/Property/lahore_multan_...          House   
2       https://www.zameen.com/Property/eden_eden_aven...          House   
3       https://www.zameen.com/Property/gulberg_2_gulb...          House   
4       https://www.zameen.com/Property/allama_iqbal_t...          House   
...                                                   ...            ...   
191388  https://www.zameen.com/Property/islamabad_i_8_...  Upper Portion   
191389  https://www.zameen.com/Property/islamabad_i_8_...  Upper Portion   
191390  https://www.zameen.com/Property/islamabad_g_10...          House   
191391  https://www.zameen.com/Property/dha_defence_dh...           Flat   
191392  https://www.zameen.com/Property/i_10_i_10_2_i_...  Upper Portion   

            price  price_bin           location       city      province_name  \
0       220000000  Very High         Model Town     Lahore             Punjab   
1        40000000  Very High        Multan Road     Lahore             Punjab   
2         9500000        Low               Eden     Lahore             Punjab   
3       125000000  Very High            Gulberg     Lahore             Punjab   
4        21000000       High  Allama Iqbal Town     Lahore             Punjab   
...           ...        ...                ...        ...                ...   
191388      70000  Very High                I-8  Islamabad  Islamabad Capital   
191389      40000     Medium                I-8  Islamabad  Islamabad Capital   
191390     160000       High               G-10  Islamabad  Islamabad Capital   
191391      25000        Low        DHA Defence  Islamabad  Islamabad Capital   
191392      26000        Low               I-10  Islamabad  Islamabad Capital   

                                         locality  ...  area_marla  area_sqft  \
0                      Model Town, Lahore, Punjab  ...       120.0   32670.12   
1                     Multan Road, Lahore, Punjab  ...        20.0    5445.02   
2                            Eden, Lahore, Punjab  ...         9.0    2450.26   
3                         Gulberg, Lahore, Punjab  ...        20.0    5445.02   
4               Allama Iqbal Town, Lahore, Punjab  ...        11.0    2994.76   
...                                           ...  ...         ...        ...   
191388          I-8, Islamabad, Islamabad Capital  ...        12.4    3375.91   
191389          I-8, Islamabad, Islamabad Capital  ...        12.4    3375.91   
191390         G-10, Islamabad, Islamabad Capital  ...        20.0    5445.02   
191391  DHA Defence, Islamabad, Islamabad Capital  ...         2.7     735.08   
191392         I-10, Islamabad, Islamabad Capital  ...         0.0       0.00   

         purpose bedrooms  date_added  year month  day  \
0       For Sale        0  07-17-2019  2019     7   17   
1       For Sale        5  10-06-2018  2018    10    6   
2       For Sale        3  07-03-2019  2019     7    3   
3       For Sale        8  04-04-2019  2019     4    4   
4       For Sale        6  04-04-2019  2019     4    4   
...          ...      ...         ...   ...   ...  ...   
191388  For Rent        3  07-24-2019  2019     7   24   
191389  For Rent        2  07-24-2019  2019     7   24   
191390  For Rent        6  07-24-2019  2019     7   24   
191391  For Rent        2  07-24-2019  2019     7   24   
191392  For Rent        3  07-24-2019  2019     7   24   

                   

# Cleaning our df

In [6]:
#OUR Required city is Lahore 
df1=df[df['city']=='Lahore']
df1

property_id  location_id  \
0            347795            8   
1            482892           48   
2            555962           75   
3            562843         3821   
4            686990         3522   
...             ...          ...   
191248     17468968         1587   
191249     17468985         1602   
191250     17469039         1607   
191251     17469096         1591   
191252     17469103         1784   

                                                 page_url  property_type  \
0       https://www.zameen.com/Property/lahore_model_t...          House   
1       https://www.zameen.com/Property/lahore_multan_...          House   
2       https://www.zameen.com/Property/eden_eden_aven...          House   
3       https://www.zameen.com/Property/gulberg_2_gulb...          House   
4       https://www.zameen.com/Property/allama_iqbal_t...          House   
...                                                   ...            ...   
191248  https://www.zameen.com/Property/dha_phase_3_dh...  Upper Portion   
191249  https://www.zameen.com/Property/dha_phase_5_dh...          House   
191250  https://www.zameen.com/Property/dha_phase_5_dh...          House   
191251  https://www.zameen.com/Property/dha_phase_4_dh...  Upper Portion   
191252  https://www.zameen.com/Property/bahria_town_ba...          House   

            price  price_bin           location    city province_name  \
0       220000000  Very High         Model Town  Lahore        Punjab   
1        40000000  Very High        Multan Road  Lahore        Punjab   
2         9500000        Low               Eden  Lahore        Punjab   
3       125000000  Very High            Gulberg  Lahore        Punjab   
4        21000000       High  Allama Iqbal Town  Lahore        Punjab   
...           ...        ...                ...     ...           ...   
191248     125000  Very High        DHA Defence  Lahore        Punjab   
191249     185000  Very High        DHA Defence  Lahore        Punjab   
191250     150000       High        DHA Defence  Lahore        Punjab   
191251      70000  Very High        DHA Defence  Lahore        Punjab   
191252      70000     Medium        Bahria Town  Lahore        Punjab   

                                 locality  ...  area_marla  area_sqft  \
0              Model Town, Lahore, Punjab  ...       120.0   32670.12   
1             Multan Road, Lahore, Punjab  ...        20.0    5445.02   
2                    Eden, Lahore, Punjab  ...         9.0    2450.26   
3                 Gulberg, Lahore, Punjab  ...        20.0    5445.02   
4       Allama Iqbal Town, Lahore, Punjab  ...        11.0    2994.76   
...                                   ...  ...         ...        ...   
191248        DHA Defence, Lahore, Punjab  ...        40.0   10890.04   
191249        DHA Defence, Lahore, Punjab  ...        20.0    5445.02   
191250        DHA Defence, Lahore, Punjab  ...        10.0    2722.51   
191251        DHA Defence, Lahore, Punjab  ...        20.0    5445.02   
191252        Bahria Town, Lahore, Punjab  ...         5.0    1361.25   

         purpose bedrooms  date_added  year month  day  \
0       For Sale        0  07-17-2019  2019     7   17   
1       For Sale        5  10-06-2018  2018    10    6   
2       For Sale        3  07-03-2019  2019     7    3   
3       For Sale        8  04-04-2019  2019     4    4   
4       For Sale        6  04-04-2019  2019     4    4   
...          ...      ...         ...   ...   ...  ...   
191248  For Rent        4  07-24-2019  2019     7   24   
191249  For Rent        5  07-24-2019  2019     7   24   
191250  For Rent        4  07-24-2019  2019     7   24   
191251  For Rent        3  07-24-2019  2019     7   24   
191252  For Rent        3  07-24-2019  2019     7   24   

                                       agency                           agent  
0                      Real Biz International                      Usama Khan  
1                                 Khan Estate      

In [7]:
required_columns=['location','area_sqft','bedrooms','price']
required_columns

['location', 'area_sqft', 'bedrooms', 'price']

In [8]:
df2=df1[required_columns]
df2

location  area_sqft  bedrooms      price
0              Model Town   32670.12         0  220000000
1             Multan Road    5445.02         5   40000000
2                    Eden    2450.26         3    9500000
3                 Gulberg    5445.02         8  125000000
4       Allama Iqbal Town    2994.76         6   21000000
...                   ...        ...       ...        ...
191248        DHA Defence   10890.04         4     125000
191249        DHA Defence    5445.02         5     185000
191250        DHA Defence    2722.51         4     150000
191251        DHA Defence    5445.02         3      70000
191252        Bahria Town    1361.25         3      70000

[58736 rows x 4 columns]

# Handling Missing Values

In [9]:
df2.isnull().sum()

location     0
area_sqft    0
bedrooms     0
price        0
dtype: int64

# Feature engineering one by one

In [10]:
df2.describe(include='all')

location     area_sqft      bedrooms         price
count         58736  5.873600e+04  58736.000000  5.873600e+04
unique          645           NaN           NaN           NaN
top     DHA Defence           NaN           NaN           NaN
freq          11208           NaN           NaN           NaN
mean            NaN  4.511614e+03      3.693238  1.944486e+07
std             NaN  1.454062e+05      1.598537  3.457283e+07
min             NaN  0.000000e+00      0.000000  1.000000e+00
25%             NaN  1.361250e+03      3.000000  2.900000e+06
50%             NaN  2.722510e+03      4.000000  1.150000e+07
75%             NaN  5.445020e+03      5.000000  2.350000e+07
max             NaN  2.450259e+07     25.000000  2.000000e+09

In [11]:
df2['bedrooms'].unique()

array([ 0,  5,  3,  8,  6,  4,  2,  7,  1, 10, 11,  9, 14, 12, 13, 18, 15,
       16, 25, 20], dtype=int64)

In [74]:
#Lets say we do not require houses having more that 8 bedrooms
df2[(df2['bedrooms']<=1) | (df2['bedrooms']>=9)]

location  area_sqft  bedrooms       price
0                     Model Town   32670.12         0   220000000
14                   Bahria Town    2722.51         0    18500000
38      Valencia Housing Society   10890.04         0    65000000
39                       Gulberg   43560.16         0  1250000000
41      Valencia Housing Society   10890.04         0    66500000
...                          ...        ...       ...         ...
191132               DHA Defence    2722.51         1       20000
191138               DHA Defence    5445.02         0      100000
191195           Sabzazar Scheme     952.88         1       20000
191226               DHA Defence    2722.51         0      130000
191241           New Muslim Town   10890.04         0      600000

[4861 rows x 4 columns]

In [75]:
df3=df2[(df2['bedrooms']>1) & (df2['bedrooms']<9)]
df3

location  area_sqft  bedrooms      price
1             Multan Road    5445.02         5   40000000
2                    Eden    2450.26         3    9500000
3                 Gulberg    5445.02         8  125000000
4       Allama Iqbal Town    2994.76         6   21000000
5                 Gulberg    5445.02         5   52000000
...                   ...        ...       ...        ...
191248        DHA Defence   10890.04         4     125000
191249        DHA Defence    5445.02         5     185000
191250        DHA Defence    2722.51         4     150000
191251        DHA Defence    5445.02         3      70000
191252        Bahria Town    1361.25         3      70000

[53875 rows x 4 columns]

# Removing area_sqft and bedrooms outliers

In [76]:
df3['area_sqft'].unique()

array([5.4450200e+03, 2.4502600e+03, 2.9947600e+03, 2.1780100e+03,
       8.7120300e+03, 2.0418800e+03, 6.5340200e+03, 1.3612500e+03,
       2.7225100e+03, 4.9005200e+03, 3.2670100e+03, 8.1675000e+02,
       3.5392630e+04, 6.8063000e+02, 5.9895200e+03, 1.6335100e+03,
       2.5591590e+04, 2.6408300e+03, 4.6282700e+03, 1.0890000e+03,
       3.8115100e+03, 1.0890040e+04, 1.4973800e+03, 5.4450000e+02,
       2.2324600e+03, 5.4450200e+04, 1.9057600e+03, 2.7497400e+03,
       4.0837600e+03, 9.2565300e+03, 3.5392600e+03, 7.6230300e+03,
       7.0785300e+03, 1.1706800e+03, 1.1979040e+04, 1.7424060e+04,
       1.2523500e+03, 7.3508000e+02, 9.8010400e+03, 1.6335060e+04,
       2.6680600e+03, 3.1036600e+03, 9.5288000e+02, 1.7696300e+03,
       1.9329800e+03, 4.1382200e+03, 3.2670120e+04, 4.0838000e+02,
       1.4157100e+03, 1.2251300e+03, 1.1162300e+03, 1.9874300e+03,
       2.9403100e+03, 1.1434500e+03, 2.1235600e+03, 5.9895000e+02,
       4.8460700e+03, 1.0345540e+04, 1.8240800e+03, 8.9843000e

In [77]:
df3['area_sqft'].dtypes

dtype('float64')

In [78]:
sqft_per_bedroom=300
df3[df3['area_sqft']/df2['bedrooms']<sqft_per_bedroom]

location  area_sqft  bedrooms    price
23              Chungi Amar Sadhu     816.75         3  3200000
35                     Karim Park     680.63         3  5500000
71              Allama Iqbal Town     544.50         3  6000000
72              Allama Iqbal Town     544.50         3  5500000
119                Al Faisal Town     544.50         3  4500000
...                           ...        ...       ...      ...
191028                Harbanspura    1361.25         5    32000
191030  Canal Bank Housing Scheme     544.50         3    16000
191186            Sabzazar Scheme    1361.25         5    50000
191192            Sabzazar Scheme     952.88         4    23000
191238          Allama Iqbal Town     816.75         3    55000

[7462 rows x 4 columns]

In [79]:
df4=df3[df3['area_sqft']/df2['bedrooms']>sqft_per_bedroom]
df4

location  area_sqft  bedrooms      price
1             Multan Road    5445.02         5   40000000
2                    Eden    2450.26         3    9500000
3                 Gulberg    5445.02         8  125000000
4       Allama Iqbal Town    2994.76         6   21000000
5                 Gulberg    5445.02         5   52000000
...                   ...        ...       ...        ...
191248        DHA Defence   10890.04         4     125000
191249        DHA Defence    5445.02         5     185000
191250        DHA Defence    2722.51         4     150000
191251        DHA Defence    5445.02         3      70000
191252        Bahria Town    1361.25         3      70000

[46413 rows x 4 columns]

# Removing location outliers

In [80]:
df4['location'].unique()

array(['Multan Road', 'Eden', 'Gulberg', 'Allama Iqbal Town',
       'Military Accounts Housing Society', 'EME Society', 'Izmir Town',
       'Upper Mall', 'Park View Villas', 'Cavalry Ground', 'Askari',
       'Bedian Road', 'Agrics Town', 'Chaman Park', 'GOR', 'Cantt',
       'Gulshan-e-Ravi', 'Al-Raheem Garden', 'Mateen Avenue',
       'Paragon City', 'IEP Engineers Town', 'Bahria Town',
       'Architects Engineers Housing Society', 'DHA Defence', 'Samanabad',
       'Mozang', 'Model Town', 'Sultan Pura',
       'T & T Aabpara Housing Society', 'Canal Garden',
       'Muslim Nagar Housing Scheme', 'Tricon Village', 'Al Faisal Town',
       'PCSIR Housing Scheme', 'Pak Arab Housing Society',
       'State Life Housing Society', 'DHA 11 Rahbar', 'Jaranwala Road',
       'Cricketer Villas', 'Rehman Villas', 'Faisal Town',
       'Harbanspura Road', 'Punjab Govt Employees Society',
       'Ashiana-e-Quaid Housing Scheme', 'Shalimar Housing Scheme',
       'Habibullah Road', 'Punjab Coo

In [81]:
len(df4['location'].unique())

574

In [82]:
df4.location=df4.location.apply(lambda x: x.strip())   #Removing all the spaces from our locations
df5=df4.groupby(df4.location)
location_count=df5['location'].agg("count").sort_values(ascending=False)
location_count

location
DHA Defence               10759
Bahria Town                5905
Johar Town                 2318
Askari                     1591
Allama Iqbal Town          1259
                          ...  
Rehman Housing Society        1
Defence Fort                  1
Rehman City Phase 6           1
Bahar Shah Road               1
Jewa Town                     1
Name: location, Length: 574, dtype: int64

In [83]:
len(location_count[location_count<=10])

351

In [84]:
location_count_lessthan_11=location_count[location_count<=10]
location_count_lessthan_11

location
Green Park                10
Dubai Town                10
Madina Colony             10
Badami Bagh               10
Ring Road                 10
                          ..
Rehman Housing Society     1
Defence Fort               1
Rehman City Phase 6        1
Bahar Shah Road            1
Jewa Town                  1
Name: location, Length: 351, dtype: int64

In [85]:
df6=df4.copy()
df6

location  area_sqft  bedrooms      price
1             Multan Road    5445.02         5   40000000
2                    Eden    2450.26         3    9500000
3                 Gulberg    5445.02         8  125000000
4       Allama Iqbal Town    2994.76         6   21000000
5                 Gulberg    5445.02         5   52000000
...                   ...        ...       ...        ...
191248        DHA Defence   10890.04         4     125000
191249        DHA Defence    5445.02         5     185000
191250        DHA Defence    2722.51         4     150000
191251        DHA Defence    5445.02         3      70000
191252        Bahria Town    1361.25         3      70000

[46413 rows x 4 columns]

In [86]:
df6['location']=df6['location'].apply(lambda x: 'other' if x in location_count_lessthan_11 else x)
df6

location  area_sqft  bedrooms      price
1             Multan Road    5445.02         5   40000000
2                    Eden    2450.26         3    9500000
3                 Gulberg    5445.02         8  125000000
4       Allama Iqbal Town    2994.76         6   21000000
5                 Gulberg    5445.02         5   52000000
...                   ...        ...       ...        ...
191248        DHA Defence   10890.04         4     125000
191249        DHA Defence    5445.02         5     185000
191250        DHA Defence    2722.51         4     150000
191251        DHA Defence    5445.02         3      70000
191252        Bahria Town    1361.25         3      70000

[46413 rows x 4 columns]

In [87]:
df6['location'].unique()

array(['Multan Road', 'Eden', 'Gulberg', 'Allama Iqbal Town',
       'Military Accounts Housing Society', 'EME Society', 'Izmir Town',
       'Upper Mall', 'Park View Villas', 'Cavalry Ground', 'Askari',
       'Bedian Road', 'other', 'Cantt', 'Gulshan-e-Ravi', 'Mateen Avenue',
       'Paragon City', 'IEP Engineers Town', 'Bahria Town',
       'Architects Engineers Housing Society', 'DHA Defence', 'Samanabad',
       'Mozang', 'Model Town', 'T & T Aabpara Housing Society',
       'Canal Garden', 'Tricon Village', 'Al Faisal Town',
       'PCSIR Housing Scheme', 'Pak Arab Housing Society',
       'State Life Housing Society', 'DHA 11 Rahbar', 'Cricketer Villas',
       'Faisal Town', 'Harbanspura Road', 'Punjab Govt Employees Society',
       'Habibullah Road', 'Punjab Coop Housing Society', 'College Road',
       'Garden Town', 'Garrison Homes', 'Khayaban-e-Amin', 'Shahdara',
       'Iqbal Park', 'Sabzazar Scheme', 'Mustafa Town', 'Walton Road',
       'New Super Town', 'West Wood Hous

In [88]:
len(df6.location.unique())

224

# Removing price outliers

In [89]:
#Creating extra column for price_per_sqft

In [90]:
df7=df6.copy()
df7

location  area_sqft  bedrooms      price
1             Multan Road    5445.02         5   40000000
2                    Eden    2450.26         3    9500000
3                 Gulberg    5445.02         8  125000000
4       Allama Iqbal Town    2994.76         6   21000000
5                 Gulberg    5445.02         5   52000000
...                   ...        ...       ...        ...
191248        DHA Defence   10890.04         4     125000
191249        DHA Defence    5445.02         5     185000
191250        DHA Defence    2722.51         4     150000
191251        DHA Defence    5445.02         3      70000
191252        Bahria Town    1361.25         3      70000

[46413 rows x 4 columns]

In [91]:
df7['pp_sqft']=df7.price/df7.area_sqft
df7

location  area_sqft  bedrooms      price       pp_sqft
1             Multan Road    5445.02         5   40000000   7346.162181
2                    Eden    2450.26         3    9500000   3877.139569
3                 Gulberg    5445.02         8  125000000  22956.756816
4       Allama Iqbal Town    2994.76         6   21000000   7012.248060
5                 Gulberg    5445.02         5   52000000   9550.010836
...                   ...        ...       ...        ...           ...
191248        DHA Defence   10890.04         4     125000     11.478378
191249        DHA Defence    5445.02         5     185000     33.976000
191250        DHA Defence    2722.51         4     150000     55.096216
191251        DHA Defence    5445.02         3      70000     12.855784
191252        Bahria Town    1361.25         3      70000     51.423324

[46413 rows x 5 columns]

In [92]:
df7['pp_sqft'].describe()

count     46413.000000
mean       6060.112500
std        4338.402653
min           0.000735
25%        1836.540545
50%        6978.854072
75%        8723.599633
max      106519.742883
Name: pp_sqft, dtype: float64

Price_per_sqft can never be 0.000735 .so removing such kind of values

In [93]:
def remove_pps_outliers(df):
    out_df=pd.DataFrame()
    
    for location,locational_df in df.groupby('location'):
        m=np.mean(locational_df['pp_sqft'])
        std=np.std(locational_df['pp_sqft'])
        reduced_df=locational_df[ (locational_df['pp_sqft']>(m-std))&
                                  (locational_df['pp_sqft']<(m+std)) ]
        out_df=pd.concat([out_df,reduced_df],ignore_index=True)
    return out_df

In [94]:
df8 = remove_pps_outliers(df7)
df8

location  area_sqft  bedrooms     price      pp_sqft
0      AWT Army Welfare Trust    2722.51         4  11000000  4040.389200
1      AWT Army Welfare Trust    5445.02         3  19000000  3489.427036
2      AWT Army Welfare Trust    1361.25         3   5000000  3673.094582
3      AWT Army Welfare Trust    5445.02         5  21000000  3856.735145
4      AWT Army Welfare Trust    2722.51         2   5600000  2056.925411
...                       ...        ...       ...       ...          ...
31209                   other    1361.25         4   7000000  5142.332415
31210                   other    1361.25         4  11000000  8080.808081
31211                   other    1361.25         4  11000000  8080.808081
31212                   other    1633.51         4  11000000  6733.965510
31213                   other    1633.51         4  12000000  7346.144193

[31214 rows x 5 columns]

In [95]:
df9=df8.drop('pp_sqft',axis='columns')
df9

location  area_sqft  bedrooms     price
0      AWT Army Welfare Trust    2722.51         4  11000000
1      AWT Army Welfare Trust    5445.02         3  19000000
2      AWT Army Welfare Trust    1361.25         3   5000000
3      AWT Army Welfare Trust    5445.02         5  21000000
4      AWT Army Welfare Trust    2722.51         2   5600000
...                       ...        ...       ...       ...
31209                   other    1361.25         4   7000000
31210                   other    1361.25         4  11000000
31211                   other    1361.25         4  11000000
31212                   other    1633.51         4  11000000
31213                   other    1633.51         4  12000000

[31214 rows x 4 columns]

# Preparing our df for model fitting

In [96]:
dummies_df=pd.get_dummies(df9['location'])
dummies_df

AWT Army Welfare Trust  Aashiana Road  \
0                           1              0   
1                           1              0   
2                           1              0   
3                           1              0   
4                           1              0   
...                       ...            ...   
31209                       0              0   
31210                       0              0   
31211                       0              0   
31212                       0              0   
31213                       0              0   

       Abdalians Cooperative Housing Society  Acro Green - Bankers Avenue  \
0                                          0                            0   
1                                          0                            0   
2                                          0                            0   
3                                          0                            0   
4                                          0                            0   
...                                      ...                          ...   
31209                                      0                            0   
31210                                      0                            0   
31211                                      0                            0   
31212                                      0                            0   
31213                                      0                            0   

       Airline Housing Society  Airport Road  Aitchison Society  \
0                            0             0                  0   
1                            0             0                  0   
2                            0             0                  0   
3                            0             0                  0   
4                            0             0                  0   
...                        ...           ...                ...   
31209                        0             0                  0   
31210                        0             0                  0   
31211                        0             0                  0   
31212                        0             0                  0   
31213                        0             0                  0   

       Al Faisal Town  Al Hafeez Gardens  Al Hamad Colony  ...  \
0                   0                  0                0  ...   
1                   0                  0                0  ...   
2                   0                  0                0  ...   
3                   0                  0                0  ...   
4                   0                  0                0  ...   
...               ...                ...              ...  ...   
31209               0                  0                0  ...   
31210               0                  0                0  ...   
31211               0                  0                0  ...   
31212               0                  0                0  ...   
31213               0                  0                0  ...   

       Valencia Housing Society  Venus Housing Scheme  Wahdat Road  \
0                             0                     0            0   
1                             0                     0            0   
2                             0                     0            0   
3                             0                     0            0   
4                             0                     0            0   
...                         ...                   ...          ...   
31209                         0                     0            0   
31210                         0                     0            0   
31211                         0                     0            0   
31212                         0                     0            0   
31213                         0                     0            0   

       Walton Road  Wapda Town  Wassanpura  West Wood Housing Soci

In [97]:
dummies_df=dummies_df.drop('other',axis='columns')
dummies_df

AWT Army Welfare Trust  Aashiana Road  \
0                           1              0   
1                           1              0   
2                           1              0   
3                           1              0   
4                           1              0   
...                       ...            ...   
31209                       0              0   
31210                       0              0   
31211                       0              0   
31212                       0              0   
31213                       0              0   

       Abdalians Cooperative Housing Society  Acro Green - Bankers Avenue  \
0                                          0                            0   
1                                          0                            0   
2                                          0                            0   
3                                          0                            0   
4                                          0                            0   
...                                      ...                          ...   
31209                                      0                            0   
31210                                      0                            0   
31211                                      0                            0   
31212                                      0                            0   
31213                                      0                            0   

       Airline Housing Society  Airport Road  Aitchison Society  \
0                            0             0                  0   
1                            0             0                  0   
2                            0             0                  0   
3                            0             0                  0   
4                            0             0                  0   
...                        ...           ...                ...   
31209                        0             0                  0   
31210                        0             0                  0   
31211                        0             0                  0   
31212                        0             0                  0   
31213                        0             0                  0   

       Al Faisal Town  Al Hafeez Gardens  Al Hamad Colony  ...  Upper Mall  \
0                   0                  0                0  ...           0   
1                   0                  0                0  ...           0   
2                   0                  0                0  ...           0   
3                   0                  0                0  ...           0   
4                   0                  0                0  ...           0   
...               ...                ...              ...  ...         ...   
31209               0                  0                0  ...           0   
31210               0                  0                0  ...           0   
31211               0                  0                0  ...           0   
31212               0                  0                0  ...           0   
31213               0                  0                0  ...           0   

       Valencia Housing Society  Venus Housing Scheme  Wahdat Road  \
0                             0                     0            0   
1                             0                     0            0   
2                             0                     0            0   
3                             0                     0            0   
4                             0                     0            0   
...                         ...                   ...          ...   
31209                         0                     0            0   
31210                         0                     0            0   
31211                         0                     0            0   
31212                         0                     0          

In [98]:
df10=pd.concat([df9,dummies_df],axis='columns')
df10

location  area_sqft  bedrooms     price  \
0      AWT Army Welfare Trust    2722.51         4  11000000   
1      AWT Army Welfare Trust    5445.02         3  19000000   
2      AWT Army Welfare Trust    1361.25         3   5000000   
3      AWT Army Welfare Trust    5445.02         5  21000000   
4      AWT Army Welfare Trust    2722.51         2   5600000   
...                       ...        ...       ...       ...   
31209                   other    1361.25         4   7000000   
31210                   other    1361.25         4  11000000   
31211                   other    1361.25         4  11000000   
31212                   other    1633.51         4  11000000   
31213                   other    1633.51         4  12000000   

       AWT Army Welfare Trust  Aashiana Road  \
0                           1              0   
1                           1              0   
2                           1              0   
3                           1              0   
4                           1              0   
...                       ...            ...   
31209                       0              0   
31210                       0              0   
31211                       0              0   
31212                       0              0   
31213                       0              0   

       Abdalians Cooperative Housing Society  Acro Green - Bankers Avenue  \
0                                          0                            0   
1                                          0                            0   
2                                          0                            0   
3                                          0                            0   
4                                          0                            0   
...                                      ...                          ...   
31209                                      0                            0   
31210                                      0                            0   
31211                                      0                            0   
31212                                      0                            0   
31213                                      0                            0   

       Airline Housing Society  Airport Road  ...  Upper Mall  \
0                            0             0  ...           0   
1                            0             0  ...           0   
2                            0             0  ...           0   
3                            0             0  ...           0   
4                            0             0  ...           0   
...                        ...           ...  ...         ...   
31209                        0             0  ...           0   
31210                        0             0  ...           0   
31211                        0             0  ...           0   
31212                        0             0  ...           0   
31213                        0             0  ...           0   

       Valencia Housing Society  Venus Housing Scheme  Wahdat Road  \
0                             0                     0            0   
1                             0                     0            0   
2                             0                     0            0   
3                             0                     0            0   
4                             0                     0            0   
...                         ...                   ...          ...   
31209                         0                     0            0   
31210                         0                     0            0   
31211                         0                     0            0   
31212                         0                     0            0   
31213                         0                     0            0   

       Walton Road  Wapda Town  Wassanpura  West Wood Housing Society  \
0                0           0           0             

In [99]:
df11=df10.drop('location',axis='columns')
df11

area_sqft  bedrooms     price  AWT Army Welfare Trust  Aashiana Road  \
0        2722.51         4  11000000                       1              0   
1        5445.02         3  19000000                       1              0   
2        1361.25         3   5000000                       1              0   
3        5445.02         5  21000000                       1              0   
4        2722.51         2   5600000                       1              0   
...          ...       ...       ...                     ...            ...   
31209    1361.25         4   7000000                       0              0   
31210    1361.25         4  11000000                       0              0   
31211    1361.25         4  11000000                       0              0   
31212    1633.51         4  11000000                       0              0   
31213    1633.51         4  12000000                       0              0   

       Abdalians Cooperative Housing Society  Acro Green - Bankers Avenue  \
0                                          0                            0   
1                                          0                            0   
2                                          0                            0   
3                                          0                            0   
4                                          0                            0   
...                                      ...                          ...   
31209                                      0                            0   
31210                                      0                            0   
31211                                      0                            0   
31212                                      0                            0   
31213                                      0                            0   

       Airline Housing Society  Airport Road  Aitchison Society  ...  \
0                            0             0                  0  ...   
1                            0             0                  0  ...   
2                            0             0                  0  ...   
3                            0             0                  0  ...   
4                            0             0                  0  ...   
...                        ...           ...                ...  ...   
31209                        0             0                  0  ...   
31210                        0             0                  0  ...   
31211                        0             0                  0  ...   
31212                        0             0                  0  ...   
31213                        0             0                  0  ...   

       Upper Mall  Valencia Housing Society  Venus Housing Scheme  \
0               0                         0                     0   
1               0                         0                     0   
2               0                         0                     0   
3               0                         0                     0   
4               0                         0                     0   
...           ...                       ...                   ...   
31209           0                         0                     0   
31210           0                         0                     0   
31211           0                         0                     0   
31212           0                         0                     0   
31213           0                         0                     0   

       Wahdat Road  Walton Road  Wapda Town  Wassanpura  \
0                0            0           0           0   
1                0            0           0           0   
2                0            0           0           0   
3                0            0           0           0   
4                0            0           0           0   
...            ...          ...         ...         ...   
31209            0        

In [100]:
X=df11.drop('price',axis='columns')
X

area_sqft  bedrooms  AWT Army Welfare Trust  Aashiana Road  \
0        2722.51         4                       1              0   
1        5445.02         3                       1              0   
2        1361.25         3                       1              0   
3        5445.02         5                       1              0   
4        2722.51         2                       1              0   
...          ...       ...                     ...            ...   
31209    1361.25         4                       0              0   
31210    1361.25         4                       0              0   
31211    1361.25         4                       0              0   
31212    1633.51         4                       0              0   
31213    1633.51         4                       0              0   

       Abdalians Cooperative Housing Society  Acro Green - Bankers Avenue  \
0                                          0                            0   
1                                          0                            0   
2                                          0                            0   
3                                          0                            0   
4                                          0                            0   
...                                      ...                          ...   
31209                                      0                            0   
31210                                      0                            0   
31211                                      0                            0   
31212                                      0                            0   
31213                                      0                            0   

       Airline Housing Society  Airport Road  Aitchison Society  \
0                            0             0                  0   
1                            0             0                  0   
2                            0             0                  0   
3                            0             0                  0   
4                            0             0                  0   
...                        ...           ...                ...   
31209                        0             0                  0   
31210                        0             0                  0   
31211                        0             0                  0   
31212                        0             0                  0   
31213                        0             0                  0   

       Al Faisal Town  ...  Upper Mall  Valencia Housing Society  \
0                   0  ...           0                         0   
1                   0  ...           0                         0   
2                   0  ...           0                         0   
3                   0  ...           0                         0   
4                   0  ...           0                         0   
...               ...  ...         ...                       ...   
31209               0  ...           0                         0   
31210               0  ...           0                         0   
31211               0  ...           0                         0   
31212               0  ...           0                         0   
31213               0  ...           0                         0   

       Venus Housing Scheme  Wahdat Road  Walton Road  Wapda Town  Wassanpura  \
0                         0            0            0           0           0   
1                         0            0            0           0           0   
2                         0            0            0           0           0   
3                         0            0            0           0           0   
4                         0            0            0           0           0   
...                     ...          ...          ...         ...         ...   
31209                     0            0            0           

In [101]:
y=df11['price']
y

0        11000000
1        19000000
2         5000000
3        21000000
4         5600000
           ...   
31209     7000000
31210    11000000
31211    11000000
31212    11000000
31213    12000000
Name: price, Length: 31214, dtype: int64

# Using LinearRegression Model

In [102]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

In [103]:
X_train,X_test,y_train,y_test=train_test_split(X,y,train_size=0.7,random_state=10)

In [104]:
lr_model=LinearRegression()

In [105]:
lr_model.fit(X_train,y_train)

LinearRegression()

In [106]:
lr_model.score(X_test,y_test)

0.6265006189018109

In [107]:
lr_model.score(X_train,y_train)

0.5737358503390448

# choosing best fit model using Grid search CV

In [108]:
from sklearn import linear_model

In [109]:
model_parameters={
     'Linear_regression':
     {
         "clf":LinearRegression(),'parameters':{'normalize':[False]}
     },
     'Linear_model.Lasso':
     {
         "clf":linear_model.Lasso(),"parameters":{'alpha':[1.0],"selection" : ['cyclic', 'random']}
     },
    'Linear_model.Ridge':
     {
         "clf":linear_model.Ridge(),"parameters":{'alpha':[1.0],
                                                  "solver" : ['auto', 'svd' ]}
     },
}

In [110]:
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import ShuffleSplit

In [111]:
cv=ShuffleSplit(n_splits=5,test_size=0.3,random_state=10)

In [112]:
scores=[]
for model_name,mp in model_parameters.items():
    clf=GridSearchCV(mp['clf'],mp['parameters'],cv=cv,return_train_score=False)
    clf.fit(X_train,y_train)
    scores.append({
        'Model':model_name,'Best parametrs':clf.best_params_,'Best_score':clf.best_score_
    })


In [115]:
scores_df=pd.DataFrame(scores)
scores_df

Model                         Best parametrs  Best_score
0   Linear_regression                   {'normalize': False}    0.506691
1  Linear_model.Lasso  {'alpha': 1.0, 'selection': 'cyclic'}    0.506691
2  Linear_model.Ridge       {'alpha': 1.0, 'solver': 'auto'}    0.506956

# Choosing LinearReession to be my final model

In [116]:
def predict_price(sqft_area,bedrooms,location):
    index_loc=np.where(X.columns==location)[0][0]
    x=np.zeros(len(X.columns))
    x[0]=sqft_area
    x[1]=bedrooms
    if index_loc>=0:
        x[index_loc]=1
    return lr_model.predict([x])[0]

In [126]:
predict_price(1633.51,4,'Zaitoon - New Lahore City')

14159818.051513758

In [118]:
X

area_sqft  bedrooms  AWT Army Welfare Trust  Aashiana Road  \
0        2722.51         4                       1              0   
1        5445.02         3                       1              0   
2        1361.25         3                       1              0   
3        5445.02         5                       1              0   
4        2722.51         2                       1              0   
...          ...       ...                     ...            ...   
31209    1361.25         4                       0              0   
31210    1361.25         4                       0              0   
31211    1361.25         4                       0              0   
31212    1633.51         4                       0              0   
31213    1633.51         4                       0              0   

       Abdalians Cooperative Housing Society  Acro Green - Bankers Avenue  \
0                                          0                            0   
1                                          0                            0   
2                                          0                            0   
3                                          0                            0   
4                                          0                            0   
...                                      ...                          ...   
31209                                      0                            0   
31210                                      0                            0   
31211                                      0                            0   
31212                                      0                            0   
31213                                      0                            0   

       Airline Housing Society  Airport Road  Aitchison Society  \
0                            0             0                  0   
1                            0             0                  0   
2                            0             0                  0   
3                            0             0                  0   
4                            0             0                  0   
...                        ...           ...                ...   
31209                        0             0                  0   
31210                        0             0                  0   
31211                        0             0                  0   
31212                        0             0                  0   
31213                        0             0                  0   

       Al Faisal Town  ...  Upper Mall  Valencia Housing Society  \
0                   0  ...           0                         0   
1                   0  ...           0                         0   
2                   0  ...           0                         0   
3                   0  ...           0                         0   
4                   0  ...           0                         0   
...               ...  ...         ...                       ...   
31209               0  ...           0                         0   
31210               0  ...           0                         0   
31211               0  ...           0                         0   
31212               0  ...           0                         0   
31213               0  ...           0                         0   

       Venus Housing Scheme  Wahdat Road  Walton Road  Wapda Town  Wassanpura  \
0                         0            0            0           0           0   
1                         0            0            0           0           0   
2                         0            0            0           0           0   
3                         0            0            0           0           0   
4                         0            0            0           0           0   
...                     ...          ...          ...         ...         ...   
31209                     0            0            0           

In [127]:
y

0        11000000
1        19000000
2         5000000
3        21000000
4         5600000
           ...   
31209     7000000
31210    11000000
31211    11000000
31212    11000000
31213    12000000
Name: price, Length: 31214, dtype: int64

##Saving our model as Lahore_Home_Price_Prediction

In [128]:
import pickle

In [129]:
with open('Lahore_Home_Price_Prediction','wb')as f:
    pickle.dump(lr_model,f)

In [130]:
import json

In [131]:
columns={
    'data_columns':[col.lower() for col in X.columns]
}
with open('columns.json','w')as f:
    f.write(json.dumps(columns))